src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,523 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 intelliGo Corporation. All rights reserved.
+//
+// Author: Fu-Yun TSUO <fy.tsuo@intelli-go.com>
+
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/igo_nr.h>
+#include <user/trace.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/igo_nr/igo_nr_comp.h>
+#include <sof/trace/trace.h>
+#include <sof/ut.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_igo_nr;
+
+uint8_t igo_heap[IGO_MEM_SIZE];
```

Comment:
We've tried to allocate 40KB on heap but failed. So we put this working buffer to bss.

Besides, this component is not expected to be instantiated more than once because it takes 250MCPS but TGL get only 400MHz.

Reply:
Hmm @lgirdwood thoughts? If this is needed at least then we should add a mechanism to prevent second instantiation and error out

Reply:
@fuyuntsuo how did you try an allocate ? This does make it a singleton and prevents us optimising the physical memory usage (i.e to a certain bank or memory type). This bss is also non static and non aligned (for cache).

Reply:
I tried the following two approaches:

1. rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, IGO_MEM_SIZE) ==> DSP failed to add widget, reply 0
2. rzalloc(SOF_MEM_ZONE_BUFFER, 0, SOF_MEM_CAPS_RAM, IGO_MEM_SIZE); ==> DSP panic, out of memory

From memory.h it seems that HEAP_BUFFER_SIZE has 0x80000 = 524,288 Bytes in total. But I still can't allocate 40KB in this region.

Reply:
@fuyuntsuo can you try rballoc()
```c
/**
 * Similar to rmalloc(), guarantees that returned block is zeroed.
 *
 * @note Do not use  for buffers (SOF_MEM_ZONE_BUFFER zone).
 *       rballoc(), rballoc_align() to allocate memory for buffers.
 */
void *rzalloc(enum mem_zone zone, uint32_t flags, uint32_t caps, size_t bytes);

/**
 * Allocates memory block from SOF_MEM_ZONE_BUFFER.
 * @param flags Flags, see SOF_MEM_FLAG_...
 * @param caps Capabilities, see SOF_MEM_CAPS_...
 * @param bytes Size in bytes.
 * @param alignment Alignment in bytes.
 * @return Pointer to the allocated memory or NULL if failed.
 */
void *rballoc_align(uint32_t flags, uint32_t caps, size_t bytes,
		    uint32_t alignment);

/**
 * Similar to rballoc_align(), returns buffer aligned to PLATFORM_DCACHE_ALIGN.
 */
static inline void *rballoc(uint32_t flags, uint32_t caps, size_t bytes)
{
	return rballoc_align(flags, caps, bytes, PLATFORM_DCACHE_ALIGN);
}
```

Reply:
+1, `rballoc` in general permits allocating larger blocks of memory, which don't _have_ to be buffers. `rmalloc`/`rzalloc` allocates small blocks of memory which are useful for various simple data structures.

Reply:
thanks! @lgirdwood @paulstelian97 @cujomalainey it works with `rballoc`.

Reply:
Here is another issue to discuss: 

In addition to the working buffer `igo_heap`, in static library we still have around 6xx KB global data storing NN coefficient. Typically we keep  these coefficient in external storage, like flash, and then DMA to RAM when the component initializes.

However there is no such external storage on TGL, so we have no choice but keep them in global data region. Do you have any suggestion?

Reply:
Constant data should be fine in the data/rodata (put a const in front of the array). Just ensure that it's only there when you have the kconfig enabled.

Reply:
That's for sure. The subdirectory won't be included when the feature is disabled in Kconfig.

Reply:
@fuyuntsuo there is existing tooling and example code for coefficient loading at runtime. Look at SRC, IIR, FIR, EQ etc for examples. @singalsu is best person to ask questions here on loading coefficients.

Reply:
@lgirdwood if you are referring to the run-time tuning tool, a.k.a. sof-ctl, it's indeed helpful for coefficient at the level of tens or hundreds of bytes. However the coefficient in this component is more than 600KB. If applying the same mechanism, it will obsoletely impose heavy burden to IPC during component prepare; moreover, it might starve IPC from other components. I am not sure whether this is a good idea.

Reply:
@fuyuntsuo got you - 600kB is quite large for IPC loading. So two options (depends if runtime tuning from host is needed).

1) Build in the coefficient at build time (ok if we dont need to tune and modify them from host at runtime).
2) create an extra source pipeline and buffer into your component and connect it with topology. That way an ALSA compressed PCM could be used to send coefficient data at runtime from tuning tools.

Reply:
IPC is pretty fast, have you tried how long pass of 600 kB via binary control takes? I haven't had problems to send some kilobytes of data for e.g. beamformer. The speed would also increase due to less overhead if the IPC size 384 bytes is increased.

Reply:
@singalsu please try an update this, IIRC there is no hard limit here why we can take up more of the mailbox.

Reply:
I agree. Depend on the physical clock rate of IPC, transmitting hundreds of KB upon component prepare might be feasible. 

But this is a bit out of the scope of this PR. If 600KB IPC passing is supported, we can treat the host driver as external storage and keep anything (text, rodata, etc.) there until SoF firmware needs them. It will be a huge change of framework imo. Currently we just need the coefficient to be linked statically according the Kconfig setting :)

Reply:
@lgirdwood we would go for 1 because NN coefficients are fixed; run-time tuning is not expected during inference.